presto-native-execution/presto_cpp/main/PrestoExchangeSource.cpp

```diff
@@ -291,10 +291,19 @@ void PrestoExchangeSource::processDataResponse(
     }
   }
 
-  int64_t ackSequence =
-      atol(headers->getHeaders()
-               .getSingleOrEmpty(protocol::PRESTO_PAGE_NEXT_TOKEN_HEADER)
-               .c_str());
+  std::optional<int64_t> ackSequenceOpt;
+  const auto nextTokenStr = headers->getHeaders().getSingleOrEmpty(
+      protocol::PRESTO_PAGE_NEXT_TOKEN_HEADER);
+  if (!nextTokenStr.empty()) {
+    // NOTE: when get data size from Presto coordinator, it might not set next
+    // token so we shouldn't update 'sequence_' if it is empty. Otherwise,
+    // 'sequence_' gets reset and we can't fetch any data from the source with
+    // the rolled back 'sequence_'.
+    ackSequenceOpt = atol(nextTokenStr.c_str());
+  } else {
+    VELOX_CHECK_EQ(
+        contentLength, 0, "next token is not set non-empty data response");
```

Comment:
... is not set for non-empty ...

Reply:
By having this check, we are not supporting data fetching from java?

Reply:
We support and only java won't set next token str in get data size request with empty data.